opacus/privacy_engine.py

```diff
@@ -350,15 +350,22 @@ def dp_step(self, closure=None, is_empty=False):
             if hasattr(self.privacy_engine.module, "ddp_hooks"):
                 # We just update the accountant
                 self.privacy_engine.steps += 1
-
+                return self.original_step(closure)
             else:
+                loss = None
+                if closure is not None:
+                    with torch.enable_grad():
+                        loss = closure()
+
```

Comment:
noob: Why isn't this necessary when module has `ddp_hooks`? Why can't this be common for both if and else blocks?

Reply:
Good question. It is worth explaining. 

First, wth is the closure for. From [PyTorch docs](https://pytorch.org/docs/stable/optim.html):

> Some optimization algorithms such as Conjugate Gradient and LBFGS need to reevaluate the function multiple times, so you have to pass in a closure that allows them to recompute your model. The closure should clear the gradients, compute the loss, and return it.
> 
> Example:
> ```python
>     def closure():
>         optimizer.zero_grad()
>         output = model(input)
>         loss = loss_fn(output, target)
>         loss.backward()
>         return loss
> ```

Before clipping and adding noise in `self.privacy_engine.step(is_empty)` we need to compute the model. Otherwise there is nothing to apply clipping/noise to. 

In case of `ddp_hooks` we just call `self.original_step(closure)` and pass the closure to call it at the original oprimiser's discretion. As the comment suggests, the clipping and noising are performed by the hooks at the end of the backward pass.

Reply:
I see. Thanks @romovpa . Can you perhaps add the above explanation as a comment to the code. (Unless, it is obvious for everyone.)